trustpinning/trustpin.go

```diff
@@ -0,0 +1,97 @@
+package trustpinning
+
+import (
+	"crypto/x509"
+	"fmt"
+	"github.com/docker/notary"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/utils"
+	"strings"
+)
+
+// CertChecker is a function type that will be used to check leaf certs against pinned trust
+type CertChecker func(leafCert *x509.Certificate, intCerts []*x509.Certificate) bool
+
+// NewTrustPinChecker returns a new certChecker function from a TrustPinConfig for a GUN
+func NewTrustPinChecker(trustPinConfig notary.TrustPinConfig, gun string) (CertChecker, error) {
+	// Determine the mode, and if it's even valid
+	if pinnedCerts, ok := trustPinConfig.Certs[gun]; ok {
+		return certsCheck(pinnedCerts), nil
+	}
+
+	if caFilepath, err := getPinnedCAFilepathByPrefix(gun, trustPinConfig); err == nil {
+		// Try to add the CA certs from its bundle file to our certificate store,
+		// and use it to validate certs in the root.json later
+		caCerts, err := trustmanager.LoadCertBundleFromFile(caFilepath)
+		if err != nil {
+			return nil, fmt.Errorf("could not load root cert from CA path")
+		}
+		// Now only consider certificates that are direct children from this CA cert, overwriting allValidCerts
+		caRootPool := x509.NewCertPool()
+		for _, caCert := range caCerts {
+			if err = trustmanager.ValidateCertificate(caCert); err != nil {
+				continue
+			}
+			caRootPool.AddCert(caCert)
+		}
+		// If we didn't have any valid CA certs, error out
+		if len(caRootPool.Subjects()) == 0 {
+			return nil, fmt.Errorf("invalid CA certs provided")
+		}
+		return caCheck(caRootPool), nil
+	}
+
+	if !trustPinConfig.DisableTOFU {
+		return tofusCheck, nil
+	}
+	return nil, fmt.Errorf("invalid trust pinning specified")
+}
+
+var certsCheck = func(pinnedCertIDs []string) CertChecker {
```

Comment:
Ah, I see what you were doing now. I was actually thinking you'd keep the struct you had before and just return a bound method:
```
type trustPinChecker struct {
    gun           string		
    config       notary.TrustPinConfig		
    pinnedCAPool *x509.CertPool
}

func NewChecker() CertChecker {
    c := trustPinChecker{}
    if pinnedCerts, ok := trustPinConfig.Certs[gun]; ok {
        return c.certsCheck, nil
    }
    ...
}
```